### ISSUE TITLE
Add missing sqlite json/jsonb functions and operators

### ISSUE THREAD

**Issue by weiznich** [diesel-rs/diesel#4366]
Diesel currently supports the sqlite json/jsonb types. We do not provide built-in support for [various methods](https://sqlite.org/draft/json1.html) available for these types. This is a tracking issue for adding support for these methods. 

The general strategy for adding support for new methods is as following:

1. Define the function via [`define_sql_function!()`](https://docs.diesel.rs/master/diesel/expression/functions/macro.define_sql_function.html). These functions can be defined [here](https://github.com/diesel-rs/diesel/tree/master/diesel/src/sqlite/expression) in a new `functions.rs` module. See the [linked definition of `to_json` for an example from the postgres backend](https://github.com/diesel-rs/diesel/blob/a55f0fef4816602889c51fc2a403cbb20563ec2b/diesel/src/pg/expression/functions.rs#L1567-L1610) as an example. This function should have a short documentation snippet with an example (See the linked sqlite documentation for examples for all of the function, please also add variants with null values, etc). If there is a `json` and a `jsonb` variant, please implement both in a single PR. 
2. If the function is generic, add a helper type definition [here](https://github.com/diesel-rs/diesel/blob/master/diesel/src/sqlite/expression/helper_types.rs). Again see the [linked definition for an example for the `to_json` function](https://github.com/diesel-rs/diesel/blob/a55f0fef4816602889c51fc2a403cbb20563ec2b/diesel/src/pg/expression/helper_types.rs#L512-L515) from the postgres backend for an example.
3. Add a test for  `#[auto_type]` support for the newly added function [here](https://github.com/diesel-rs/diesel/blob/6c1cb1d42eb2b3000502d67249dc3cdd76066dd7/diesel_derives/tests/auto_type.rs#L413)
5. Submit a PR with the change

Method list:

- [x] [`json(json)`](https://sqlite.org/draft/json1.html#jmini) *
- [x] [`jsonb(json)`](https://sqlite.org/draft/json1.html#jminib) *
- [ ] [`json_array(value1, value2, …)`](https://sqlite.org/draft/json1.html#jarray) **
- [ ] [`jsonb_array(value1, value2, …)`](https://sqlite.org/draft/json1.html#jarrayb) **
- [x] [`json_array_length(json)`](https://sqlite.org/draft/json1.html#jarraylen) *
- [x] [`json_array_length(json, path)`](https://sqlite.org/draft/json1.html#jarraylen) *
- [x] [`json_error_position(json)`](https://sqlite.org/draft/json1.html#jerr) *
- [ ] [`json_extract(json, path, …)`](https://sqlite.org/draft/json1.html#jex) **
- [ ] [`jsonb_extract(json, path, …)`](https://sqlite.org/draft/json1.html#jexb) **
- [ ] [`json_insert(json, path, value, …)`](https://sqlite.org/draft/json1.html#jins) **
- [ ] [`jsonb_insert(json, path, value, …)`](https://sqlite.org/draft/json1.html#jinsb) **
- [ ] [`json_object(label1, value1, …)`](https://sqlite.org/draft/json1.html#jobj) **
- [ ] [`jsonb_object(label1, value1, …)`](https://sqlite.org/draft/json1.html#jobjb) **
- [ ] [`json_patch(json1, json2)`](https://sqlite.org/draft/json1.html#jpatch) *
- [ ] [`jsonb_patch(json1, json2)`](https://sqlite.org/draft/json1.html#jpatchb) *
- [x] [`json_pretty(json)`](https://sqlite.org/draft/json1.html#jpretty) *
- [ ] [`json_remove(json, path, …)`](https://sqlite.org/draft/json1.html#jrm) **
- [ ] [`jsonb_remove(json, path, …)`](https://sqlite.org/draft/json1.html#jrmb) **
- [ ] [`json_replace(json,path,value,...)`](https://sqlite.org/draft/json1.html#jrepl) **
- [ ] [`jsonb_replace(json,path,value,...)`](https://sqlite.org/draft/json1.html#jreplb) **
- [ ] [`json_set(json,path,value,...)`](https://sqlite.org/draft/json1.html#jset) **
- [ ] [`jsonb_set(json,path,value,...)`](https://sqlite.org/draft/json1.html#jsetb) **
- [x] [`json_type(json)`](https://sqlite.org/draft/json1.html#jtype) *
- [x] [`json_type(json,path)`](https://sqlite.org/draft/json1.html#jtype) *
- [x] [`json_valid(json)`](https://sqlite.org/draft/json1.html#jvalid) *
- [ ] [`json_valid(json,flags)`](https://sqlite.org/draft/json1.html#jvalid) *
- [ ] [`json_quote(value)`](https://sqlite.org/draft/json1.html#jquote) *

There are four [aggregate SQL functions](https://sqlite.org/draft/lang_aggfunc.html):

- [ ] [`json_group_array(value)`](https://sqlite.org/draft/json1.html#jgrouparray) ***
- [ ] [`jsonb_group_array(value)`](https://sqlite.org/draft/json1.html#jgrouparrayb) ***
- [ ] [`json_group_object(label,value)`](https://sqlite.org/draft/json1.html#jgroupobject) ***
- [ ] [`jsonb_group_object(name,value)`](https://sqlite.org/draft/json1.html#jgroupobjectb) ***

For items marked with * the instructions above can be followed as written down

Items marked with ** are variadic functions on SQL side. Rust does not support such function definitions yet, so we either need to have variants for a certain number of fixed arguments or we need to find another solution.

For items marked with *** the function definition needs to be marked with the `#[aggregate]` attribute.

# Operators:

- [ ] [`json -> path`](https://sqlite.org/draft/json1.html#jptr)
- [ ] [`json ->>`](https://sqlite.org/draft/json1.html#jptr)

For operators:

These already exists for the postgres backend [here](https://github.com/diesel-rs/diesel/blob/a55f0fef4816602889c51fc2a403cbb20563ec2b/diesel/src/pg/expression/operators.rs#L46-L52). We need to look for a way to share these impls.

**Comment by xuehaonan27** [comment]
I would like to take `json(json)` and `jsonb(json)`

**Comment by KekmaTime** [comment]
I will work on  `json_valid(json)` & `json_type(json)`

**Comment by xuehaonan27** [comment]
May I work on more tasks?

**Comment by weiznich** [comment]
@xuehaonan27 Sure that would be very welcome

**Comment by zaira-bibi** [comment]
I'll work on `json_array_length`

**Comment by IamTossan** [comment]
Hey, I took care of `json_array_length`

**Comment by KekmaTime** [comment]
@weiznich i think you unintentionally closed the entire issue 😅

**Comment by weiznich** [comment]
Correct, that shouldn't have happened. (Github treats "Partially fixes #xyz" as command to close that issue 😞 )

**Comment by JakobKlocker** [comment]
I'd implement json_patch and jsonb_patch next

**Comment by mertwole** [comment]
I'd take `json_group_array` and `jsonb_group_array`

**Comment by mertwole** [comment]
I'll also take `json_array` and `jsonb_array`

**Comment by mertwole** [comment]
@weiznich Do you have any thoughts on how variadic functions support should be implemented? I had some thoughts on it and it seems like the most reasonable way of doing this would be to add support for the `#[variadic]` attribute to the 
`#[declare_sql_function]`. It may look like

```rust
#[variadic(repeated = (name, value), generate_up_to = 16)]
fn json_insert<J: Json, N: SqlType + SingleValue, V: SqlType + SingleValue>(json: J, name: N, value: V) -> Json;
```

which will generate 16 functions named `json_insert_1`, `json_insert_2` ..., `json_insert_16`

Pros:
- This implementation can be shared across all backends(if I'm not mistaken variadic functions for pg are also not yet implemented)
- It will not introduce boilerplate
- Another solution can easily be implemented later without re-writing much of the code

Cons:
- Compilation time and binary size will increase due to the codegen (though it can be softened by feature-gating variadic functions)
- Limited amount of arguments

**Comment by weiznich** [comment]
@mertwole I do not have a plan for that ready yet. I would like to avoid generating many more instances for these cases, so maybe we can come up with another solution.
Given that we "know" the types of the variadic arguments we could maybe just use a slice or something there to pass in the variable number of arguments? So something like

```
fn json_insert(json: J, variadic: [(N, V)]) -> Json
```

**Comment by mertwole** [comment]
@weiznich What do you mean by that we "know" these types? All we know about the argument types is that they implement `SqlType + SingleValue`, so I suppose there're only 2 ways of passing them as slices:
- pass `Box<dyn SqlType + SingleValue>`
- introduce a enum containing variants for all the types supported by `sqlite` and implement `From<SpecificType>` for it (for this variant we'll lose compile-time argument type checking)

**Comment by weiznich** [comment]
@mertwole You are correct that won't work. Trait objects won't help there and using an enum is also not possible as this is an open set of types. You could write as of today something like `json_pretty(some_column.contcat("whatever"))` and it would construct the correct SQL.
That written: I do not have better proposals.

**Comment by weiznich** [comment]
@mertwole Let's try the macro variant, but with the following modifications:

* Default to 2 variants
* Check for an env-variable: `DIESEL_VARIADIC_FUNCTION_ARGS` and if that's set to a larger number generate as many variants as requested. (That can be set in a `.cargo/config.toml` file per project). That should have some documentation.

That lets user easily request more parameters as they need while keeping the default overhead low.

**Comment by mertwole** [comment]
@weiznich Ok, you're right, this seems like a more flexible approach

**Comment by mertwole** [comment]
I'll also take `json_remove` and `jsonb_remove`

**Comment by kvasi34** [comment]
Can I take `json_extract` and `jsonb_extract`?

**Comment by weiznich** [comment]
@KonstantinosVasilopoulos That would be welcome. Please take a look at how it's handled for the `json_object` function to have an example to follow.

**Comment by kvasi34** [comment]
Thanks @weiznich !

I am not sure what data type the `json_extract` function should return. I am thinking `Nullable<Text>`, since the function can literally return anything; Json, string, NULL, integer. The problem with this approach is that an integer would be returned as its string representation (eg 7 -> "7").

Otherwise, we could go for `Nullable<Untyped>` and let the user decide? I think that `Nullable<Text>` is a good compromise though.

Edit 31 March: I am writing this here because of the "mentoring available" label. If this is not the appropriate space, kindly indicate to me where I could ask for guidance.

I tried creating a new marker trait (JsonExtractValue) to represent the return type of the `json_extract` function; the trait basically implements all possible `json_extract` return types (Integer, Text, Json, etc.). However, the `#[declare_sql_function]` macro requires every generic to correspond to an argument (JsonExtractValue is used for the return value, not for an argument) , resulting in a compile-time error. Any advice on how I should proceed?

**Comment by eyupcanakman** [comment]
I'd like to take `json_insert` and `jsonb_insert`.

`json_insert` is straightforward following `json_remove`, but `jsonb_insert` hits a snag. When SQLite inserts new text through a path, it stores the new object keys and string values as TEXTRAW elements (type `0x0A`), and the JSONB reader in `sqlite/types/json.rs` currently errors with "TEXTRAW is not supported". So `jsonb_insert` builds correct SQL, but reading the result back into `serde_json::Value` fails for any insert that adds a key or a string. The same will hit `jsonb_set` and `jsonb_replace` later, since they store inserted text the same way.

Supporting TEXTRAW looks like a small change, since the payload is just raw text that can go through the same path as TEXT. Would you prefer I include that in this PR, do it as a separate reader PR first, or leave it out and document the limitation on `jsonb_insert` for now?

**Comment by weiznich** [comment]
@eyupcanakman Thanks for your interest in working on this. I think it's reasonable to just add support for `TEXTRAW` on the side while implementing the actual functions. Using the existing `read_jsonb_text` function sounds like the right approach there.
